coccinelle/namedata.cocci

```diff
@@ -0,0 +1,11 @@
+// NameData is a fixed-size type of 64 bytes. Using strlcpy to copy data into a
+// NameData struct can cause problems because any data that follows the initial
+// null-terminated string will also be part of the data.
+
+@rule_namedata_strlcpy@
+expression E1, E2;
+symbol NAMEDATALEN;
+@@
+- strlcpy(E1, E2, NAMEDATALEN);
++ /* You are using strlcpy with NAMEDATALEN, please consider using NameData and namestrcpy instead. */
++ namestrcpy(&E1, E2);
```

Comment:
Not sure if you should have the `&` there: the expressions are already pointers, so if you take the address of a pointer this will not work in most cases.

Reply:
My assumption was that the variable is also changed from `char[]` to `NameData`. In this case, the `&` operator is required here. However, since the change of the variable type is not part of this rule and it depends on the context, I have removed the suggestion and only annotate the `strlcpy` call.

Reply:
I think after this change most likely `E1` in its original will not work anymore...the rule does drag attention to it - and I think that's what it should do!

Reply:
> My assumption was that the variable is also changed from `char[]` to `NameData`. In this case, the `&` operator is required here. 
Exactly.

> However, since the change of the variable type is not part of this rule and it depends on the context, I have removed the suggestion and only annotate the `strlcpy` call.

Well. Given that the NAMEDATALEN is part of the rule, I think it quite clear that the target is a pointer to a buffer with this length, so it makes sense to still suggest it as a transformation. Especially considering the fact that it is a `strlen`. It is quite common to use `memcpy` as well, but this will copy all the data and does not suffer from this problem.

I actually think that you can add the "struct" rule that you used before. It is quite clearly the case that when you store something like this in a structure, you should use `NameData`. Local variables is a different story.

Reply:
Good points. I added the suggestion of the transformation for `strlcpy` and the rule for struct members. In addition, I enabled Coccinelle checks for *.h files in the newest version of this PR.